Add a certificate authority to freeipa and move DNS and security group jazz here

jsf9k opened the pull request:
## 🗣 Description

This pull request contains the changes necessary to this repository to support the changes made in cisagov/freeipa-server-tf-module#25.  Specifically, the AWS Route53 DNS and VPC security group resources from [cisagov/freeipa-server-tf-module](https://github.com/cisagov/freeipa-server-tf-module) now exist here.

## 💭 Motivation and Context

[cisagov/freeipa-server-tf-module](https://github.com/cisagov/freeipa-server-tf-module) required some major changes because:
* We require a special KDC certificate in order to support PIV-only `kinit` via PKINIT.  Therefore we let FreeIPA create its own certificate authority so that it can generate the appropriate certificate automatically.  Allowing FreeIPA to function as the CA for all certificates hosted on machines joined to the domain also has the advantage that FreeIPA will handle certificate renewal automatically.
* I want to be able to use the same module for both masters and replicas.  (Once they are completely set up, masters and replicas are indistinguishable.)  Therefore it makes sense to remove the master-specific cloud-init jazz from this repo and rename it from freeipa-master-tf-module to freeipa-server-tf-module.
* Removing the FreeIPA cloud-init setup jazz also means that the admin PW and directory server PW are no longer needed as input variables.

These changes necessitated changes in this repository as well.

## 🧪 Testing

I have successfully deployed these changes (together with those in cisagov/freeipa-server-tf-module#25) to our staging environment.

## 🚥 Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (causes existing functionality to change)

## ✅ Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

jsf9k committed:
Remove cert read roles and comment out replicas

Also use a specific branch of cisagov/freeipa-master-tf-module.

jsf9k committed:
Rename output

jsf9k committed:
Update freeipa module configuration to match recent changes in that TF module

jsf9k committed:
Correct output references

jsf9k committed:
Remove variable that no longer exists

jsf9k committed:
Remove admin and directory service PWs from freeipa-server-tf-module config

They are no longer used there.

jsf9k committed:
Change module source to match renamed repo

jsf9k committed:
Add DNS code that was originally from freeipa-server-tf-module as a submodule

jsf9k committed:
Remove three variables that are no longer needed

jsf9k committed:
Update README to match recent changes

jsf9k committed:
Add security group from freeipa-server-tf-module here

This helps that repo support being used to deploy both freeipa masters
and replicas.

jsf9k committed:
Add two more IPA servers

jsf9k committed:
Remove some old code that was commented out

jsf9k committed:
Improve comment

jsf9k committed:
Add a TTL variable to use for DNS records

jsf9k committed:
Add non-master DNS records

jsf9k committed:
Rename dns to master_dns

The master IPA server requires more DNS records than does the replica
IPA server.

jsf9k committed:
Remove the tags variable

There are no resources in the master_dns module that accept tags.

jsf9k committed:
Create a replica_dns submodule to create A and PTR records for IPA replica servers

jsf9k committed:
Add TTL variable to README

jsf9k committed:
Modify freeipa.tf to use new replica_dns submodule

dav3r commented on security_groups/locals.tf:
```diff
@@ -0,0 +1,37 @@
+locals {
+  # The ports that the IPA servers should listen on
+  ipa_ports = {
```
Don't we need the DNS TCP/UDP ports in this list?

jsf9k replied:
![image](https://user-images.githubusercontent.com/5521725/83687050-38480c00-a5b9-11ea-840d-44b72ba5aec1.png)

Not yet.  I opted to stick with AWS DNS for now, since I would have to rip out the zones that are created in [cisagov/cool-sharedservices-networking](https://github.com/cisagov/cool-sharedservices-networking) in order to get FreeIPA hosting DNS.  Too many things changing at once!

I plan to revisit this later.  There are definite advantages to letting FreeIPA host DNS for hosts that are attached to the domain (automatic DNS updates as IPA servers are added and removed and as hosts join/leave the domain), but I need to verify that it won't interefere with public DNS, for example.

dav3r replied:
Got it - thanks for the explanation! 👍

jsf9k committed:
Add support for load balancing

In this commit I
* Rename the master_dns submodule to dns
* Remove the replica_dns submodule
* Rework the dns submodule to take a map of hostname keys and IP
values as an input, and from that create DNS records that result in
equal load balancing across all servers.

jsf9k committed:
Use the new dns submodule to create DNS records that load balance equally across all IPA servers

jsf9k commented:
@felddy and @dav3r, have a look at commits 45ec4b9 and 4dfa666, wherein I create DNS records that result in load balancing across all IPA servers.  This is some new :fire: :hankey: that I am excited about!

dav3r commented on dns/README.md:
```diff
@@ -0,0 +1,48 @@
+  source = "./master_dns"
+
+  domain          = "example.com"
+  hostname_ip_map = "ipa.example.com"
```
This does not appear to be a valid map for this example.

jsf9k replied:
See commit c98d9a8.

dav3r commented on dns/README.md:
```diff
@@ -0,0 +1,48 @@
+|------|-------------|------|---------|:--------:|
+| domain | The domain for the IPA master (e.g. example.com). | `string` | n/a | yes |
+| hostname_ip_map | A map whose keys are the hostnames of the IPA servers and whose values are the IPs corresponding to  those servers (e.g. {"ipa0.example.com" = "10.0.0.1", "ipa1.example.com" = "10.0.0.2"}). | `map(string)` | n/a | yes |
+| reverse_zone_id | The zone ID corresponding to the private Route53 reverse zone where the PTR records related to the IPA master should be created (e.g. ZKX36JXQ8W82L). | `string` | n/a | yes |
```
Minor nitpick, but you have the same example zone id listed here and in the `zone_id` below.  Typically, our input examples match what we have in the `Usage` section above.

jsf9k replied:
See commit c98d9a8.

dav3r commented on README.md:
```diff
@@ -37,36 +26,39 @@ module "example" {
-| replica1 | The first IPA replica EC2 instance. |
-| replica2_certificate_read_role | The IAM role used by the second IPA replica to read its certificate information. |
-| replica2 | The second IPA replica EC2 instance. |
+| server0 | The first IPA server EC2 instance. |
```
You are missing the outputs for the other two servers.

jsf9k replied:
See commit a5afae9.

dav3r commented on dns/variables.tf:
```diff
@@ -0,0 +1,37 @@
+variable "zone_id" {
+  type        = string
+  description = "The zone ID corresponding to the private Route53 zone where the Kerberos-related DNS records should be created (e.g. ZKX36JXQ8W82L)."
+}
```
Same minor nitpick as above.

jsf9k replied:
See commit 3910d11.

dav3r commented on freeipa.tf:
```diff
@@ -27,79 +27,65 @@ locals {
-  source = "github.com/cisagov/freeipa-replica-tf-module"
+# Create the IPA servers
+module "ipa0" {
+  source = "github.com/cisagov/freeipa-server-tf-module?ref=improvement%2Fadd-ca"
```
⚠️ Usual reminder about pointing to a code branch.

jsf9k replied:
See commit dd19020.

dav3r commented on freeipa.tf:
```diff
@@ -27,79 +27,65 @@ locals {
+  tags                 = merge(var.tags, map("Name", "FreeIPA 0"))
+}
+module "ipa1" {
+  source = "github.com/cisagov/freeipa-server-tf-module?ref=improvement%2Fadd-ca"
```
⚠️ Usual reminder about pointing to a code branch.

jsf9k replied:
See commit dd19020.

dav3r commented on freeipa.tf:
```diff
@@ -27,79 +27,65 @@ locals {
+  tags                 = merge(var.tags, map("Name", "FreeIPA 1"))
 }
+module "ipa2" {
+  source = "github.com/cisagov/freeipa-server-tf-module?ref=improvement%2Fadd-ca"
```
⚠️ Usual reminder about pointing to a code branch.

jsf9k replied:
See commit dd19020.

dav3r reviewed (changes requested):
This is awesome sauce, but I noted a couple of things that need to be corrected before I approve.

dav3r commented:
> @felddy and @dav3r, have a look at commits [45ec4b9](https://github.com/cisagov/cool-sharedservices-freeipa/commit/45ec4b9926d9f40ef3c0d37f7addecf939259c40) and [4dfa666](https://github.com/cisagov/cool-sharedservices-freeipa/commit/4dfa6666402b595fd5314e74c12b328c326da4f5), wherein I create DNS records that result in load balancing across all IPA servers. This is some new 🔥 💩 that I am excited about!

Super 🔥 💩 indeed!  🥇

jsf9k committed:
Add two other servers as outputs in README

jsf9k committed:
Correct example

jsf9k committed:
Make the zone ID and reverse zone IDs different IDs

This is to make it clearer to the user that they should be different
zones.

jsf9k committed:
Remove branch reference

It is no longer needed (or valid) now that
cisagov/freeipa-server-tf-module#25 has been merged.

dav3r reviewed (approved):
All good now! 🧨

jsf9k committed:
Whittle down codeowners

felddy reviewed (approved):
Looks good!  🎆 
Sorry it took so long to review this.  📆 
Just for posterity: Let's Encrypt *cannot* issue certificates with the configurations FreeIPA needs.
